sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,492 @@
+        } else {
+          return true;
+        }
+      } catch (InterruptedException e) {
+        throw new IOException(e);
+      }
+    }
+
+    @Override
+    public void close() throws IOException {
+      LOGGER.info("Closing MQTT reader");
+      try {
+        if (client != null) {
+          try {
+            client.disconnect();
+          } finally {
+            client.close();
+          }
+        }
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+    }
+
+    @Override
+    public Instant getWatermark() {
+      return watermark;
+    }
+
+    @Override
+    public UnboundedSource.CheckpointMark getCheckpointMark() {
+      return new UnboundedSource.CheckpointMark() {
+        @Override
+        public void finalizeCheckpoint() throws IOException {
+          // nothing to do
+        }
+      };
+    }
+
+    @Override
+    public byte[] getCurrent() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return current;
+    }
+
+    @Override
+    public UnboundedMqttSource getCurrentSource() {
+      return source;
+    }
+
+    @Override
+    public Instant getCurrentTimestamp() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return currentTimestamp;
+    }
+
+  }
+
+  /**
+   * A {@link PTransform} to write and send a message to a MQTT server.
+   */
+  @AutoValue
+  public abstract static class Write extends PTransform<PCollection<byte[]>, PDone> {
+
+    @Nullable abstract MqttConnectionConfiguration mqttConnectionConfiguration();
+    @Nullable abstract int qos();
+    @Nullable abstract boolean retained();
+
+    abstract Builder toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setMqttConnectionConfiguration(MqttConnectionConfiguration configuration);
+      abstract Builder setQos(int qos);
+      abstract Builder setRetained(boolean retained);
+      abstract Write build();
+    }
+
+    public Write withMqttConnectionConfiguration(MqttConnectionConfiguration configuration) {
+      return toBuilder().setMqttConnectionConfiguration(configuration).build();
+    }
+
+    public Write withQoS(int qos) {
+      return toBuilder().setQos(qos).build();
+    }
+
+    public Write withRetained(boolean retained) {
+      return toBuilder().setRetained(retained).build();
+    }
+
+    @Override
+    public PDone apply(PCollection<byte[]> input) {
+      input.apply(ParDo.of(new MqttWriter(this)));
+      return PDone.in(input.getPipeline());
+    }
+
+    @Override
+    public void validate(PCollection<byte[]> input) {
+      mqttConnectionConfiguration().validate();
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      mqttConnectionConfiguration().populateDisplayData(builder);
+      builder.add(DisplayData.item("qos", qos()));
+      builder.add(DisplayData.item("retained", retained()));
+    }
+
+    /**
```

Comment:
This comment provides zero information over what the class name already provides.